File: packages/android-compose/src/main/kotlin/dev/chartcn/mobile/ApiAdapter.kt
package dev.chartcn.mobile

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull

enum class ApiCursorLocation {
  QUERY,
  BODY
}

data class ApiPaginationConfig(
  val cursorParamName: String = "cursor",
  val nextCursorPath: String = "pagination.nextCursor",
  val maxPages: Int = 20,
  val maxRows: Int = 5000,
  val cursorLocation: ApiCursorLocation = ApiCursorLocation.QUERY
)

data class ApiPageRequest(
  val source: Source,
  val pageIndex: Int,
  val cursor: String?,
  val query: Map<String, JsonElement>,
  val body: Map<String, JsonElement>
)

data class ApiPageResponse(
  val payload: JsonElement,
  val nextCursor: String? = null
)

fun interface ApiPageFetcher {
  fun fetch(request: ApiPageRequest): ApiPageResponse
}

class ApiAdapter(private val fetcher: ApiPageFetcher) {
  fun fetchRows(
    source: Source,
    pagination: ApiPaginationConfig = ApiPaginationConfig()
  ): List<ChartRow> {
    require(source.adapter == Adapter.API) {
      "Source adapter must be 'api'."
    }
    require(!source.endpoint.isNullOrBlank()) {
      "api source requires a non-empty endpoint."
    }
    require(!source.method.isNullOrBlank()) {
      "api source requires a non-empty method."
    }
    require(pagination.maxPages > 0) {
      "maxPages must be greater than 0."
    }
    require(pagination.maxRows > 0) {
      "maxRows must be greater than 0."
    }
    require(pagination.cursorParamName.matches(Regex("^[a-zA-Z0-9_.-]+$"))) {
      "cursorParamName contains unsupported characters."
    }

    val rows = mutableListOf<ChartRow>()
    val seenCursors = mutableSetOf<String>()
    var cursor: String? = null
    var pageIndex = 0

    while (pageIndex < pagination.maxPages && rows.size < pagination.maxRows) {
      val query = when (pagination.cursorLocation) {
        ApiCursorLocation.QUERY -> withCursor(source.query, cursor, pagination.cursorParamName)
        ApiCursorLocation.BODY -> source.query
      }
      val body = when (pagination.cursorLocation) {
        ApiCursorLocation.QUERY -> source.body
        ApiCursorLocation.BODY -> withCursor(source.body, cursor, pagination.cursorParamName)
      }

      val response = fetcher.fetch(
        ApiPageRequest(
          source = source,
          pageIndex = pageIndex,
          cursor = cursor,
          query = query,
          body = body
        )
      )

      val pageRows = extractRows(response.payload, source.dataPath)
      if (pageRows.isNotEmpty()) {
        val remaining = pagination.maxRows - rows.size
        rows.addAll(pageRows.take(remaining))
      }

      val nextCursor = response.nextCursor?.takeIf { it.isNotBlank() }
        ?: extractStringAtPath(response.payload, pagination.nextCursorPath)

      if (nextCursor == null || pageRows.isEmpty()) {
        break
      }

      if (!seenCursors.add(nextCursor)) {
        break
      }

      cursor = nextCursor
      pageIndex += 1
    }

    return rows
  }
}

private fun withCursor(
  base: Map<String, JsonElement>,
  cursor: String?,
  paramName: String
): Map<String, JsonElement> {
  if (cursor.isNullOrBlank()) return base
  return base + (paramName to JsonPrimitive(cursor))
}

private fun extractRows(
  payload: JsonElement,
  dataPath: String?
): List<ChartRow> {
  val target = payload.valueAtPath(dataPath)
  return when (target) {
    is JsonArray -> target.mapNotNull { entry ->
      (entry as? JsonObject)?.toMap()
    }
    is JsonObject -> listOf(target.toMap())
    else -> emptyList()
  }
}

private fun extractStringAtPath(
  payload: JsonElement,
  path: String
): String? {
  val value = payload.valueAtPath(path) as? JsonPrimitive ?: return null
  return value.contentOrNull?.takeIf { it.isNotBlank() }
}

private fun JsonElement?.valueAtPath(path: String?): JsonElement? {
  if (this == null) return null
  if (path.isNullOrBlank()) return this

  val parts = path.split(".").filter { it.isNotBlank() }
  if (parts.isEmpty()) return this

  var current: JsonElement = this
  for (part in parts) {
    current = when (current) {
      is JsonObject -> current[part] ?: return null
      is JsonArray -> {
        val index = part.toIntOrNull() ?: return null
        current.getOrNull(index) ?: return null
      }
      else -> return null
    }
  }
  return current
}

File: packages/ios-swiftui/Sources/ChartCNMobile/APIAdapter.swift
import Foundation

public enum APICursorLocation: Sendable {
    case query
    case body
}

public struct APIPaginationOptions: Sendable {
    public let cursorParameter: String
    public let nextCursorPath: String
    public let maxPages: Int
    public let maxRows: Int
    public let cursorLocation: APICursorLocation

    public init(
        cursorParameter: String = "cursor",
        nextCursorPath: String = "pagination.nextCursor",
        maxPages: Int = 20,
        maxRows: Int = 5000,
        cursorLocation: APICursorLocation = .query
    ) {
        self.cursorParameter = cursorParameter
        self.nextCursorPath = nextCursorPath
        self.maxPages = maxPages
        self.maxRows = maxRows
        self.cursorLocation = cursorLocation
    }
}

public struct APIPageRequest: Sendable {
    public let source: ChartSpec.Source
    public let pageIndex: Int
    public let cursor: String?
    public let query: [String: ChartValue]
    public let body: [String: ChartValue]
}

public struct APIPageResponse: Sendable {
    public let payload: ChartValue
    public let nextCursor: String?

    public init(payload: ChartValue, nextCursor: String? = nil) {
        self.payload = payload
        self.nextCursor = nextCursor
    }
}

public typealias APIPageFetcher = @Sendable (APIPageRequest) async throws -> APIPageResponse

public final class APIAdapter {
    private let fetchPage: APIPageFetcher

    public init(fetchPage: @escaping APIPageFetcher) {
        self.fetchPage = fetchPage
    }

    public func fetchRows(
        from source: ChartSpec.Source,
        pagination: APIPaginationOptions = .init()
    ) async throws -> [ChartRow] {
        guard source.adapter == .api else {
            throw ChartRuntimeError.invalidSpec("Source adapter must be 'api'.")
        }
        guard let endpoint = source.endpoint?.trimmingCharacters(in: .whitespacesAndNewlines), !endpoint.isEmpty else {
            throw ChartRuntimeError.invalidSpec("api source requires a non-empty endpoint.")
        }
        guard let method = source.method?.trimmingCharacters(in: .whitespacesAndNewlines), !method.isEmpty else {
            throw ChartRuntimeError.invalidSpec("api source requires a non-empty method.")
        }
        guard pagination.maxPages > 0 else {
            throw ChartRuntimeError.invalidSpec("maxPages must be greater than 0.")
        }
        guard pagination.maxRows > 0 else {
            throw ChartRuntimeError.invalidSpec("maxRows must be greater than 0.")
        }
        guard pagination.cursorParameter.range(of: "^[a-zA-Z0-9_.-]+$", options: .regularExpression) != nil else {
            throw ChartRuntimeError.invalidSpec("cursorParameter contains unsupported characters.")
        }

        var rows: [ChartRow] = []
        var seenCursors = Set<String>()
        var cursor: String?
        var pageIndex = 0

        while pageIndex < pagination.maxPages, rows.count < pagination.maxRows {
            let baseQuery = source.query ?? [:]
            let baseBody = source.body ?? [:]
            let query: [String: ChartValue]
            let body: [String: ChartValue]

            switch pagination.cursorLocation {
            case .query:
                query = withCursor(base: baseQuery, cursor: cursor, parameter: pagination.cursorParameter)
                body = baseBody
            case .body:
                query = baseQuery
                body = withCursor(base: baseBody, cursor: cursor, parameter: pagination.cursorParameter)
            }

            let request = APIPageRequest(
                source: source.with(endpoint: endpoint, method: method),
                pageIndex: pageIndex,
                cursor: cursor,
                query: query,
                body: body
            )
            let response = try await fetchPage(request)
            let pageRows = extractRows(from: response.payload, dataPath: source.dataPath)

            if !pageRows.isEmpty {
                let remaining = max(0, pagination.maxRows - rows.count)
                rows.append(contentsOf: pageRows.prefix(remaining))
            }

            let nextCursor = normalizedCursor(response.nextCursor)
                ?? extractCursor(from: response.payload, path: pagination.nextCursorPath)

            if nextCursor == nil || pageRows.isEmpty {
                break
            }

            guard let unwrappedCursor = nextCursor, seenCursors.insert(unwrappedCursor).inserted else {
                break
            }

            cursor = unwrappedCursor
            pageIndex += 1
        }

        return rows
    }
}

private func withCursor(
    base: [String: ChartValue],
    cursor: String?,
    parameter: String
) -> [String: ChartValue] {
    guard let cursor = normalizedCursor(cursor) else {
        return base
    }
    var next = base
    next[parameter] = .string(cursor)
    return next
}

private func extractRows(
    from payload: ChartValue,
    dataPath: String?
) -> [ChartRow] {
    guard let target = value(atPath: dataPath, in: payload) else {
        return []
    }

    switch target {
    case .array(let values):
        return values.compactMap { value in
            guard case .object(let row) = value else { return nil }
            return row
        }
    case .object(let row):
        return [row]
    default:
        return []
    }
}

private func extractCursor(
    from payload: ChartValue,
    path: String
) -> String? {
    guard let value = value(atPath: path, in: payload) else {
        return nil
    }

    switch value {
    case .string(let text):
        return normalizedCursor(text)
    case .number(let number):
        return normalizedCursor(String(number))
    case .bool(let flag):
        return normalizedCursor(String(flag))
    default:
        return nil
    }
}

private func value(atPath path: String?, in root: ChartValue) -> ChartValue? {
    guard let path, !path.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        return root
    }

    var current: ChartValue = root
    let parts = path.split(separator: ".").map(String.init).filter { !$0.isEmpty }

    for part in parts {
        switch current {
        case .object(let object):
            guard let next = object[part] else {
                return nil
            }
            current = next
        case .array(let values):
            guard let index = Int(part), values.indices.contains(index) else {
                return nil
            }
            current = values[index]
        default:
            return nil
        }
    }

    return current
}

private func normalizedCursor(_ cursor: String?) -> String? {
    guard let value = cursor?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else {
        return nil
    }
    return value
}

private extension ChartSpec.Source {
    func with(endpoint: String, method: String) -> ChartSpec.Source {
        .init(
            adapter: adapter,
            entity: entity,
            predicate: predicate,
            sort: sort,
            limit: limit,
            table: table,
            where: `where`,
            args: args,
            orderBy: orderBy,
            queryName: queryName,
            endpoint: endpoint,
            method: method,
            headers: headers,
            query: query,
            body: body,
            dataPath: dataPath,
            rows: rows
        )
    }
}
